### PR TITLE
patch-rs binding

### PR DESCRIPTION
**Pull request by juchiast**
I'm using git dependency for patch-rs, because v0.2 is not published yet.

### CODE REVIEW

File: src/bindings/app/patch.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+use rlua::{Lua, UserData, UserDataMethods};
+
+struct PatchParser(patch_rs::PatchParser);
+
+impl UserData for PatchParser {
+    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
+        methods.add_method("process", |_, this, _: ()| {
+            Ok(this.0.process().expect("patch_parser error"))
```
Comment (dariusc93) [inline src/bindings/app/patch.rs]: Shouldnt use `expect`. Should include `rlua::prelude::LuaError`, remove the `Ok` and do `this.0.process().map_err(LuaError::external)` that way if there is an error, it wont panic.

**Comment by juchiast** [comment]
I tried that, but the compiler complains:

```
error[E0277]: `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
 --> src/bindings/app/patch.rs:8:38
  |
8 |             this.0.process().map_err(rlua::Error::external)
  |                                      ^^^^^^^^^^^^^^^^^^^^^ `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
  |
  = help: the trait `std::marker::Sync` is not implemented for `(dyn std::error::Error + std::marker::Send + 'static)`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(dyn std::error::Error + std::marker::Send + 'static)>`
  = note: required because it appears within the type `std::boxed::Box<(dyn std::error::Error + std::marker::Send + 'static)>`
  = note: required because it appears within the type `std::option::Option<std::boxed::Box<(dyn std::error::Error + std::marker::Send + 'static)>>`
  = note: required because it appears within the type `error_chain::State`
  = note: required because it appears within the type `patch_rs::PatchError`
  = note: required because of the requirements on the impl of `failure::Fail` for `patch_rs::PatchError`
  = note: required because of the requirements on the impl of `std::convert::From<patch_rs::PatchError>` for `failure::Error`
  = note: required because of the requirements on the impl of `std::convert::Into<failure::Error>` for `patch_rs::PatchError`
  = note: required by `rlua::Error::external`
```

How about `this.0.process().ok()`? The downside is that we lost the error message.

**Comment by juchiast** [comment]
@dariusc93

I tried that, but the compiler complains:

```
error[E0277]: `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
 --> src/bindings/app/patch.rs:8:38
  |
8 |             this.0.process().map_err(rlua::Error::external)
  |                                      ^^^^^^^^^^^^^^^^^^^^^ `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
  |
  = help: the trait `std::marker::Sync` is not implemented for `(dyn std::error::Error + std::marker::Send + 'static)`
  = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(dyn std::error::Error + std::marker::Send + 'static)>`
  = note: required because it appears within the type `std::boxed::Box<(dyn std::error::Error + std::marker::Send + 'static)>`
  = note: required because it appears within the type `std::option::Option<std::boxed::Box<(dyn std::error::Error + std::marker::Send + 'static)>>`
  = note: required because it appears within the type `error_chain::State`
  = note: required because it appears within the type `patch_rs::PatchError`
  = note: required because of the requirements on the impl of `failure::Fail` for `patch_rs::PatchError`
  = note: required because of the requirements on the impl of `std::convert::From<patch_rs::PatchError>` for `failure::Error`
  = note: required because of the requirements on the impl of `std::convert::Into<failure::Error>` for `patch_rs::PatchError`
  = note: required by `rlua::Error::external`
```

How about `this.0.process().ok()`? The downside is that we lost the error message.

**Comment by juchiast** [comment]
error-chain is not compatible with failure: https://github.com/rust-lang-nursery/error-chain/issues/240

I'm doing this:
```
            this.0
                .process()
                .map_err(|e| rlua::Error::external(failure::err_msg(e.to_string())))

```

**Comment by juchiast** [comment]
@dariusc93 
Ok, I tested current master branch of patch-rs, it works!
Publish a new version to crates.io then I'll update this PR.

**Comment by hedgar2017** [comment]
@juchiast done, v0.3.0 is out there.
@dariusc93 I agree, I should have used `failure` from the beginning. Some of my projects still use `error-chain` so for some reason I went for that.

**Comment by mitchtbaum** [comment]
@hedgar2017, :+1: :+1: 
@juchiast :+1: :+1: 
@dariusc93 :+1: :+1:

**Pull request merged**